main/src/library/File.flix

```diff
@@ -840,4 +840,100 @@ namespace File {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.list(): Array[String] & Impure;
+
+            let javaFile = newFile(f);
+            let javaList = list(javaFile);
+
+            Ok(Array.toList(javaList))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an Iterator naming the files and directories in the directory `f`.
+    ///
+    /// Does not recursively traverse the directory.
+    ///
+    pub def listWithIter(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.newDirectoryStream(##java.nio.file.Path): ##java.nio.file.DirectoryStream & Impure;
+            import java.nio.file.DirectoryStream.iterator(): ##java.util.Iterator & Impure;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = newDirectoryStream(javaPath);
+            let javaIter = iterator(javaStream);
+            convertPathIter(javaIter, fullPath = false)
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an Iterator naming the files and directories in the directory `f`.
+    ///
+    /// Recursively traverses the directory.
+    ///
+    pub def walk(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.walk(##java.nio.file.Path, Array[##java.nio.file.FileVisitOption]): ##java.util.stream.Stream & Impure;
+            import java.util.stream.BaseStream.iterator(): ##java.util.Iterator & Impure;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = walk(javaPath, []);
+            let javaIter = iterator(javaStream as ##java.util.stream.BaseStream);
+            convertPathIter(javaIter, fullPath = true)
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Converts a java iterator of Paths to a Flix Iterator of Strings.
+    ///
+    /// If `opt.fullPath` is `true`, then the whole path is specified. Otherwise only the name.
+    ///
+    def convertPathIter(iter: ##java.util.Iterator, opt: {fullPath :: Bool}): Result[Iterator[String], String] & Impure =
```

Comment:
I am not sure about the name of this helper method

Reply:
Not sure about the name, but I was thinking this helper function would not use `fullPath`. Is that possible? 

Could we write a generic reusable iterator2iterator?

Reply:
+1 for general Java Iterator to Flix Iterator.

Reply:
We could also consider something like `def catchIoException(block: () -> a & ef): a & ef` to deduplicate all the try-catches. Might obscure things a bit though, so I'm not sure.